### ISSUE TITLE
Multicanvas API Thoughts

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#2338]
## 🚀 Feature

This comment right now is really a collection of some of my thoughts around the multicanvas API 

--------

### Background context and minor questions

There are a number of issue discussing multiple canvases that already exist in napari - #561, #662, and #760, including discussion of our grid mode. Most of them though are a little out of date, and we've made some good progress since then, doing thinks like adding `viewer.camera` in the mean time.

We're also making some progress towards bringing slicing off the layers, see some discussion in #1353 around our layer models too, and adding layer groups too.

The big question I have now is what functionality should we support and what should the API look like.

One thing I'll assert is that I think a `Layer` should NOT know what canvas it is in, and making this change should not require changes to our layer models (as much as possible), curious if others disagree.

There should be some way to assign layers (or layer groups) to canvases AND it should be possible to assign one layer to multiple canvases (this would allow you to easily reuse the same layer as a "background" in multiple canvases without copying the layer), see request in https://github.com/napari/napari/issues/561#issuecomment-788535358.

**Question** If one layer is assigned to multiple canvases should all of it's attributes then be forced to be the same for all canvases, even if they aren't "spatially varying" for example colormap? If no then I think we're at much more of a "data source" like approach as discussed at places in #1775 where all the spatially varying data is combined into a "data source". My preference is yes as it makes everything much simpler, I think if you want "no" I'd probably say you want to just add another copy of the layer, but I maybe missing something and if the answer is no, then we'll need to think differently as we'll need to have multiple layers sharing the same datasource (and just one layer per canvas), instead of multiple canvases sharing the same layer and no real concept of an independent datasource.

One thing that I think is clear though is that a layer in one canvas should be able to be sliced independently of that same layer in another canvas. Assuming we went with a "yes" to the above, all it's other properties should be the same. This would enable things like orthoviews but with global control over clims etc. If we went with "no" above then we'd need to use functionality like was provided in #2226 to link them, which could be ok too.

**Question** how do we relate our current single dims model to the different slices of a layer? Do we have one `Dims` that has all the information it needs, or does each canvas have it's own Dims? There was some good discussion of a global dims in https://github.com/napari/napari/issues/760#issuecomment-565739856.

Another thing we need to consider is how to represent the camera or multiple cameras. In general we can have one global camera (like grid mode) or one camera per canvas, and which can be linked in the sense that zooming one zooms the other (or a more complex linkage), or unlinked in the sense that they can be controlled independently. **Question** how do we want to represent the camera/ cameras?

**Question** Does each canvas exist in the same world coordinates space, or can they be totally unrelated/ independent?

------------------------------------------------------------

### Main Question

Stepping back, right now we have a single `viewer.layers` layer list, a single `viewer.camera`, a single `viewer.dims`, a single `viewer.cursor`, and no `viewer.canvas`, but a `viewer.grid` that controls grid mode.

My main question is what does the API for multiple canvases now look like?

Do we have `viewer.cameras` which is a list of `Camera` objects one per canvas? Or `viewer.dims` a list of `Dims` objects one per canvas? etc

Do we have `viewer.canvases` a list of canvases? With each of them having their own camera/ dims/ layers/ cursor? How do these then relate to the global dims/ camera/ cursor/ layers. Do we still have them, are they derived from them?

We could definitely start with an implementation of just a more simple special case, but i think the general discussion is worth it now as it would be good to be confident we have the right API. I'd also really prefer not to break everything now for the single canvas use case to support the multicanvas one.

I'll keep thinking about this, and then might have an update with a more concrete proposal

**Comment by sofroniewn** [comment]
Note the motivation for this post was recent discussion in https://github.com/napari/napari/issues/561#issuecomment-788535358

**Comment by JoOkuma** [comment]
From my anecdotal experience implementing [this ortho viewer plugin](https://github.com/JoOkuma/ortho-view-napari) a lot of things could be accomplished using the `(un)link_layers` functions and using different ordering in each viewer `Dims`. 

I think extending the linked layers concept might be key to some of these GitHub issues, even for other problems as slicing the data we could solve it by providing different canvas' interactions when linking different layers. For example, we could create a `Slicing` layer that changes the Vispy behavior for its links.

Displaying 3D data is currently an issue with multiple viewers, the 3D rendering is automatically shown on the latter viewer, so I think allowing multiple 3D renderings without the grid view would require some changes to the Vispy API as well.

We could list the applications we want to support with the multi canvas API because if the data between canvases are not somewhat related (linked) I don't see why it can't be implemented using the current Napari API and multiple viewers.

I'm not aware of use cases where multiple cameras are necessary, especially with the transformations API.

**Comment by brisvag** [comment]
Great to see some collection fo throughts on this, I think this is going to be a key step up in possibilities for napari! I want to point out a few vispy things that will be useful (if not necessary) for this to work:
1. there was some discussion about using vispy's `Grid`. While it might be useful for quiclky exploding a lot of layers in a grid, its implementation is quite limited if one wants to resize things. Something more intuitive to use might be achieved with normal qt splitters [like in this vispy example](https://github.com/vispy/vispy/blob/main/examples/demo/gloo/two_qt_widgets.py). I'd love to be able to just drag the edge of the first canvas and spawn a second one, where I can just change the camera view and whatnot.
2. vispy has the `VisualView`s, which (as far as I understand) are designed exactly for this purpose: showing the same data in different ways. Some of the Visuals may need some adjusting to be used this way (e.g: set up more stuff in functions like `_prepare_draw`), but should already work as intended for things like camera transforms (thanks to `_prepare_transforms` and such).

**Comment by sofroniewn** [comment]
Yeah, I'd probably like us to have a "model" for multicanvas support on the napari / pydantic side, but then maybe we use some of the vispy machinery on the view side for performance benefits
